### ISSUE TITLE
tardigrade integration tests failing - range requests not working properly

### ISSUE THREAD

**Issue by ncw** [rclone/rclone#4370]
The integration tests for [tardigrade are failing](https://pub.rclone.org/integration-tests/2020-06-18-050025/index.html) like this

![image](https://user-images.githubusercontent.com/536803/85050684-5ce1dd80-b18e-11ea-81cc-89b159467790.png)

The failures all look related and to do with Range requests failing.

The first one is easy to demonstrate

```
$ cd backend/tardigrade
$ go test -v -run TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange
=== RUN   TestIntegration
    TestIntegration: fstests.go:356: Using remote "TestTardigrade:"
=== RUN   TestIntegration/FsMkdir
=== RUN   TestIntegration/FsMkdir/FsPutFiles
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange
    TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange: fstests.go:238: 
        	Error Trace:	fstests.go:238
        	            				fstests.go:1284
        	Error:      	Received unexpected error:
        	            	ranger error: buffer runoff
        	            		storj.io/common/ranger.ByteRanger.Range:39
        	            		storj.io/uplink/private/stream.(*Download).resetReader:111
        	            		storj.io/uplink/private/stream.(*Download).Read:62
        	            		storj.io/uplink.(*Download).Read:70
        	            		bytes.(*Buffer).ReadFrom:204
        	            		io/ioutil.readAll:36
        	            		io/ioutil.ReadAll:45
        	            		github.com/rclone/rclone/fstest/fstests.readObject:237
        	            		github.com/rclone/rclone/fstest/fstests.Run.func13.13.30:1284
        	            		testing.tRunner:991
        	Test:       	TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange
        	Messages:   	readObject("file name.txt") limit=-1, options=[RangeOption(81,100000)]
2020/06/18 18:06:51 ERROR : FS sj://rclone-test-gohizob2padoboj8qadecun4: Failed to list "": directory not found
2020/06/18 18:06:51 ERROR : : Failed to rmdir: directory not found
--- FAIL: TestIntegration (2.58s)
    --- FAIL: TestIntegration/FsMkdir (2.18s)
        --- FAIL: TestIntegration/FsMkdir/FsPutFiles (1.47s)
            --- FAIL: TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange (0.18s)
FAIL
exit status 1
FAIL	github.com/rclone/rclone/backend/tardigrade	2.588s
```

The ultimate cause here seems to be the `ranger error: buffer runoff` from the tardigrade libraries

Here is the test in question

https://github.com/rclone/rclone/blob/85bcacac905be0fb08696a5035dc1ce0dc89f377/fstest/fstests/fstests.go#L1274-L1288

It is this test which fails `{fs.RangeOption{Start: 81, End: 100000}, 81, 100}` so it looks like the backend needs to clip the range at the size of the file.

Here is the code in question

https://github.com/rclone/rclone/blob/85bcacac905be0fb08696a5035dc1ce0dc89f377/backend/tardigrade/object.go#L140-L158

It looks like we need to read the size of the object for the `s && e` case too.

@calebcase I remember you didn't want to read the size of the object if you didn't have to. Do you think it is acceptable to read it here?

This will likely mean that all range requests read the size of the object, so maybe we should use the built in fs.FixRangeOption ?

Or is this something best fixed in the tardigrade libraries?

**Comment by calebcase** [comment]
This is related to a different regression in how errors are returned from storj/uplink. There's a new version of storj/uplink that has the fix we need in it. I'll make a PR soon.

**Comment by ncw** [comment]
> This is related to a different regression in how errors are returned from storj/uplink. There's a new version of storj/uplink that has the fix we need in it. I'll make a PR soon.

Thanks, I'll check it out in the morning 😃

**Comment by ncw** [comment]
Hopefully this is fixed by #4375
